### PR TITLE
fix: remove python version upper bound

### LINKED ISSUE #944: looker-sdk 21.20.0 throws TypeError for python 3.9.10
Due to a [change](https://github.com/python/cpython/blame/v3.9.10/Lib/functools.py#L855) in Python's `functools` module that was included in the update from 3.9.9->3.9.10, `looker_sdk` breaks on import. The error seems to be through the `cattr` dependency:

```
  File "/opt/app/src/tvsapi/views.py", line 8, in <module>
    import looker_sdk
  File "/usr/local/lib/python3.9/site-packages/looker_sdk/__init__.py", line 32, in <module>
    from looker_sdk.sdk.api31 import methods, methods as methods31  # noqa:F401
  File "/usr/local/lib/python3.9/site-packages/looker_sdk/sdk/api31/methods.py", line 32, in <module>
    from . import models as mdls
  File "/usr/local/lib/python3.9/site-packages/looker_sdk/sdk/api31/models.py", line 11935, in <module>
    sr.converter31.register_structure_hook(
  File "/usr/local/lib/python3.9/site-packages/cattr/converters.py", line 186, in register_structure_hook
    self._structure_func.register_cls_list([(cl, func)])
  File "/usr/local/lib/python3.9/site-packages/cattr/multistrategy_dispatch.py", line 45, in register_cls_list
    self._single_dispatch.register(cls, handler)
  File "/usr/local/lib/python3.9/functools.py", line 855, in register
      raise TypeError(
TypeError: Invalid first argument to `register()`. ForwardRef('AccessToken') is not a class.
```

### PR DESCRIPTION
**Pull request by joeldodge79**
fixes #944

the main fix here is switching from `cattr.register_structure_hook` to
`cattr.register_structure_hook_func`. I re-organized some of the code
in an effort to reproduce the issue in the test_serialize.py suite but
I could not reproduce there, only in integration tests.

Now that it's a function, we don't need a generated hook per type.

I further isolated the converter usage switching the serializing to use
each 31/40 dedicated converter.

I also switched to the newer GenConverter.

### CODE REVIEW

File: packages/sdk-codegen/src/python.gen.ts [hunk 1]
```diff
@@ -476,10 +471,6 @@ ${this.hooks.join('\n')}
       }
     }
 
-    const forwardRef = `ForwardRef("${type.name}")`
-    this.hooks.push(
-      `sr.converter${this.apiRef}.register_structure_hook(\n${bump}${forwardRef},  # type: ignore\n${bump}${this.structureHookFR}  # type:ignore\n)`
-    )
```
Comment (joeldodge79) [inline packages/sdk-codegen/src/python.gen.ts:482]: no longer need a register hook per type

File: packages/sdk-codegen/src/python.gen.ts [hunk 2]
```diff
@@ -141,18 +140,14 @@ DelimSequence = model.DelimSequence
...
-  this.structureHookFR
-} = functools.partial(sr.forward_ref_structure_hook, globals(), sr.converter${
-    this.apiRef
-  })
+forward_ref_structure_hook = functools.partial(
+    sr.forward_ref_structure_hook, globals(), sr.converter${this.apiRef}
+)
+sr.converter${this.apiRef}.register_structure_hook_func(
+    lambda t: t.__class__ is ForwardRef, forward_ref_structure_hook
+)
```
Comment (joeldodge79) [inline packages/sdk-codegen/src/python.gen.ts:150]: now it's just this one new `register_structure_hook_func`

File: python/looker_sdk/__init__.py [hunk 3]
```diff
@@ -62,7 +62,7 @@ def init31(
     return methods31.Looker31SDK(
         auth_session.AuthSession(settings, transport, serialize.deserialize31, "3.1"),
         serialize.deserialize31,
-        serialize.serialize,
+        serialize.serialize31,
```
Comment (joeldodge79) [inline python/looker_sdk/__init__.py:65]: let's keep the cattrs converters separate both on input and output

File: python/looker_sdk/rtl/serialize.py [hunk 4]
```diff
@@ -75,34 +75,12 @@ def deserialize(
...
-    """Map top level json keys to model property names.
-
-    Currently this translates reserved python keywords like "from" => "from_"
-    """
-    for reserved in keyword.kwlist:
-        if reserved in data and isinstance(data, dict):
-            data[f"{reserved}_"] = data.pop(reserved)
-    return data
-
-
```
Comment (joeldodge79) [inline python/looker_sdk/rtl/serialize.py:100]: moved to new hooks.py file

File: python/looker_sdk/rtl/serialize.py [hunk 5]
```diff
@@ -124,51 +102,30 @@ def forward_ref_structure_hook(context, converter, data, forward_ref):
...
-        return parser.isoparse(d)
-
-else:
-
-    def datetime_structure_hook(
-        d: str, t: Type[datetime.datetime]
-    ) -> datetime.datetime:
-        return datetime.datetime.strptime(d, DATETIME_FMT)
-
-
```
Comment (joeldodge79) [inline python/looker_sdk/rtl/serialize.py:170]: moved to new hooks.py file

File: python/setup.py [hunk 6]
```diff
@@ -55,7 +55,7 @@
     package_data={"looker_sdk": ["py.typed", "looker_sdk/looker-sample.ini"]},
     packages=find_packages(),
     # restrict python to <=3.9.9 due to https://github.com/looker-open-source/sdk-codegen/issues/944
-    python_requires=">=3.6, <=3.9.9",
+    python_requires=">=3.6",
```
Comment (joeldodge79) [inline python/setup.py:58]: up up and away

File: python/tests/rtl/test_serialize.py [hunk 7]
```diff
@@ -95,6 +95,7 @@ class Model(ml.Model):
     list_model_no_refs1: Sequence["ModelNoRefs1"]
     opt_enum1: Optional["Enum1"] = None
     opt_model_no_refs1: Optional["ModelNoRefs1"] = None
+    list_opt_model_no_refs1: Optional[Sequence["ModelNoRefs1"]] = None
```
Comment (joeldodge79) [inline python/tests/rtl/test_serialize.py:98]: thought I saw some odd behavior so I added this version of type to make sure cattrs gets it right

**Review by jkaster — APPROVED** [review]
LGTM. Thanks for the inline comments

File: python/setup.py (hunk 6 above)
Comment (alekseiloginov) [reply python/setup.py:58]: I think we can also delete the comment now in a followup

**Pull request merged**